aospy/automate.py

```diff
@@ -0,0 +1,247 @@
+                if isinstance(obj, type_)])
+
+
+class CalcSuite(object):
+    """Generate suite of Calc objects based on inputted specifications."""
+
+    _CORE_SPEC_NAMES = ('projects',
+                        'models',
+                        'runs')
+
+    _AUX_SPEC_NAMES = ('variables',
+                       'date_ranges',
+                       'regions',
+                       'input_time_intervals',
+                       'input_time_datatypes',
+                       'input_time_offsets',
+                       'input_vertical_datatypes',
+                       'output_time_intervals',
+                       'output_time_regional_reductions',
+                       'output_vertical_reductions')
+
+    _NAMES_SUITE_TO_CALC_MAPPING = {
+        'projects': 'proj',
+        'models': 'model',
+        'runs': 'run',
+        'variables': 'var',
+        'date_ranges': 'date_range',
+        'regions': 'region',
+        'input_time_intervals': 'intvl_in',
+        'input_time_datatypes': 'dtype_in_time',
+        'input_time_offsets': 'time_offset',
+        'input_vertical_datatypes': 'dtype_in_vert',
+        'output_time_intervals': 'intvl_out',
+        'output_time_regional_reductions': 'dtype_out_time',
+        'output_vertical_reductions': 'dtype_out_vert',
+    }
+
+    def __init__(self, calc_suite_specs, obj_lib):
+        """Turn all inputs into aospy-ready objects."""
+        self._specs_in = calc_suite_specs
+        self._obj_lib = obj_lib
+
+        self._projects = self._get_projects()
+
+    def _get_projects(self):
+        projs = _get_all_objs_of_type(Proj, self._obj_lib)
+        names = self._specs_in['projects']
+        return set([proj for proj in projs if proj.name in names])
+
+    def _get_regions(self):
+        return [_get_all_objs_of_type(Region, self._obj_lib)]
+
+    def _get_variables(self):
+        if hasattr(self._obj_lib, 'variables'):
+            return self._obj_lib.variables
+        else:
+            return _get_all_objs_of_type(Var, self._obj_lib)
+
+    def _permute_core_specs(self):
+        obj_trees = []
+        for project in self._projects:
+            models = _get_mult_objs_by_names(self._specs_in['models'], project,
+                                             'models', 'default_models')
+            for model in models:
+                runs = _get_mult_objs_by_names(self._specs_in['runs'], model,
+                                               'runs', 'default_runs')
+                for run in runs:
+                    # Use Calc's attr names as keys.
+                    obj_trees.append(dict(proj=project, model=model, run=run))
+        return obj_trees
+
+    def _get_aux_specs(self):
+        # Drop the "core" specifications, which are handled separately.
+        specs = self._specs_in.copy()
+        [specs.pop(core) for core in self._CORE_SPEC_NAMES]
+        specs['regions'] = self._get_regions()
+        specs['variables'] = self._get_variables()
```

Comment:
(Similar to above) I think this will *always* use every variable defined by the user in their object library.  Is that correct?